connectors/sources/github.py

```diff
@@ -1371,50 +1528,50 @@ async def _get_repo_object_for_github_app(self, owner, repo_name):
-                        self.foreign_repos[
-                            repo_name
-                        ] = await self.github_client.get_foreign_repo(
-                            repo_name=repo_name
-                        )
-                    except Exception:
-                        self._logger.debug(f"Detected invalid repository: {repo_name}.")
-                        invalid_repos.append(repo_name)
+                all_repos = configured_repos + foreign_repos
             else:
                 foreign_repos, configured_repos = self.github_client.bifurcate_repos(
                     repos=self.configured_repos,
                     owner=self.configuration["org_name"],
                 )
-                configured_repos.extend(foreign_repos)
-                if self.configuration["org_name"] not in self.org_repos:
-                    self.org_repos[self.configuration["org_name"]] = {}
-                async for repo in self.github_client.get_org_repos(
-                    self.configuration["org_name"]
-                ):
-                    self.org_repos[self.configuration["org_name"]][
-                        repo["nameWithOwner"]
-                    ] = repo
-                invalid_repos = list(
-                    set(configured_repos)
-                    - set(self.org_repos[self.configuration["org_name"]].keys())
+                all_repos = configured_repos + foreign_repos
+
+            invalid_repos = []
+            if all_repos:
+                batch_results = (
+                    await self.github_client.get_repos_by_fully_qualified_name_batch(
+                        all_repos
+                    )
                 )
+                for repo_name, repo_data in batch_results.items():
+                    if repo_data:
+                        # Store valid repos for potential later use
+                        if repo_name in foreign_repos:
+                            self.foreign_repos[repo_name] = repo_data
+                        else:
+                            # Store configured repos in the appropriate cache
+                            if self.configuration["repo_type"] == "other":
+                                logged_in_user = await self._logged_in_user()
+                                if logged_in_user not in self.user_repos:
+                                    self.user_repos[logged_in_user] = {}
+                                self.user_repos[logged_in_user][repo_name] = repo_data
+                            else:
+                                org_name = self.configuration["org_name"]
+                                if org_name not in self.org_repos:
+                                    self.org_repos[org_name] = {}
+                                self.org_repos[org_name][repo_name] = repo_data
+                    else:
+                        self._logger.debug(f"Detected invalid repository: {repo_name}.")
+                        invalid_repos.append(repo_name)
```

Comment:
Can we reduce the nesting by one level by capturing this at the top of the loop

```python
if not repo_data:
    invalid_repos.appen()
    logger.debug...
    continue
```

I like avoiding messing with the loop control flow but I think this is one of the cases where it's welcome, to make the rest of the code easier to reason about.

Reply:
Also, optional, these type of snippets
```python
if org_name not in self.org_repos:
    self.org_repos[org_name] = {}
self.org_repos[org_name][repo_name] = repo_data
```

can become

```python
self.org_repos.setdefault(org_name, {})[repo_name] = repo_data
```

but I don't have a strong opinion if you think the current version is more readable.